### PR TITLE
Remove JsonView filter

### PR DESCRIPTION
**Pull request by dstepanov**
This PR changes how JsonView used to work before. Instead of detecting the `JsonView` annotation using the filter and invoking serialization, the responsibility is moved to the `JsonMapper`. This requires the method annotation to be correctly propagated to the serializer; we need to copy the method annotations if the argument is unwrapped. Jax-RS also expects that.

The property to enable the view is `jackson.json-view.enabled`, but that also worked for Serde, so we need to fix it before the next minor release.

### CODE REVIEW

**Review by yawkat — APPROVED** [review]

**Comment by dstepanov** [comment]
@sdelamo Is there something wrong with the test I wrote? It looks like the configuration is not applied for GraalVM

**Comment by yawkat** [comment]
maybe the metadata is lost when running without reflection?

**Comment by dstepanov** [comment]
Looks like Jackson needs reflective metadata to read the view annotation

**Review by graemerocher — APPROVED** [review]

File: jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonMediaTypeCodec.java [hunk 1]
```diff
@@ -64,13 +65,24 @@ public JsonMediaTypeCodec(ObjectMapper objectMapper,
      * @param applicationConfiguration The common application configurations
      * @param codecConfiguration The configuration for the codec
      */
-    @Inject
     public JsonMediaTypeCodec(BeanProvider<ObjectMapper> objectMapper,
```
Comment (sdelamo) [inline jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonMediaTypeCodec.java:70]: can we deprecate this constructor?

File: jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonStreamMediaTypeCodec.java [hunk 2]
```diff
@@ -75,7 +76,6 @@ public JsonStreamMediaTypeCodec(ObjectMapper objectMapper,
      * @param applicationConfiguration The common application configurations
      * @param codecConfiguration The configuration for the codec
      */
-    @Inject
```
Comment (sdelamo) [inline jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonStreamMediaTypeCodec.java:78]: can we deprecate this constructor?

File: jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java [hunk 3]
```diff
@@ -67,19 +68,31 @@
...
 
-    @Inject
     @Internal
     public JacksonDatabindMapper(ObjectMapper objectMapper) {
+        this(objectMapper, false);
+    }
+
+    @Inject
+    @Internal
+    public JacksonDatabindMapper(ObjectMapper objectMapper, @Value("${" + JacksonDatabindMapper.PROPERTY_JSON_VIEW_ENABLED + ":false}") boolean allowViews) {
```
Comment (sdelamo) [inline jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java:93]: should not we default to true?

File: jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java [hunk 4]
```diff
@@ -67,19 +68,31 @@
...
     private final ObjectWriter specializedWriter;
+    private final boolean allowViews;
 
     private TypeCache<ObjectReader> cachedReader;
     private TypeCache<ObjectWriter> cachedWriter;
 
-    @Inject
     @Internal
     public JacksonDatabindMapper(ObjectMapper objectMapper) {
+        this(objectMapper, false);
```
Comment (sdelamo) [inline jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java:88]: should not we default to `true`?

File: jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java (hunk 3 above)
Comment (sdelamo) [reply jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java:93]: never mind. I see it was defaulting to false already.

File: jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java (hunk 4 above)
Comment (sdelamo) [reply jackson-databind/src/main/java/io/micronaut/jackson/databind/JacksonDatabindMapper.java:88]: never mind. I see it was defaulting to false already.

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewCodecResolver.java [hunk 5]
```diff
@@ -1,34 +0,0 @@
...
-import io.micronaut.core.annotation.NonNull;
-import io.micronaut.json.codec.JsonMediaTypeCodec;
-
-/**
- * Interface for resolving codecs for {@link com.fasterxml.jackson.annotation.JsonView} types.
- *
- * @author graemerocher
- * @since 1.1
- */
-interface JsonViewCodecResolver {
```
Comment (sdelamo) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewCodecResolver.java:27]: removing this is a breaking change

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewMediaTypeCodecFactory.java [hunk 6]
```diff
@@ -1,66 +0,0 @@
...
- *
- * @since 1.1
- * @author mmindenhall
- * @author graemerocher
- */
-@Requires(beans = JsonConfiguration.class)
-@Requires(property = JsonViewServerFilter.PROPERTY_JSON_VIEW_ENABLED)
-@Singleton
-@Primary
-class JsonViewMediaTypeCodecFactory implements JsonViewCodecResolver {
```
Comment (sdelamo) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewMediaTypeCodecFactory.java:41]: removing this is a breaking change

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewCodecResolver.java (hunk 5 above)
Comment (yawkat) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewCodecResolver.java:27]: the classes are package private

File: jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonStreamMediaTypeCodec.java (hunk 2 above)
Comment (sdelamo) [reply jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonStreamMediaTypeCodec.java:78]: see: https://github.com/micronaut-projects/micronaut-core/pull/11194

File: jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonMediaTypeCodec.java (hunk 1 above)
Comment (sdelamo) [reply jackson-databind/src/main/java/io/micronaut/jackson/codec/JsonMediaTypeCodec.java:70]: see: https://github.com/micronaut-projects/micronaut-core/pull/11194

**Review by sdelamo — APPROVED** [review]
Please, add the deprecation notices https://github.com/micronaut-projects/micronaut-core/pull/11194 even if the classes are themselves already deprecated.

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java [hunk 7]
```diff
@@ -1,121 +0,0 @@
...
- * @since 1.1
- * @author mmindenhall
- * @author graemerocher
- */
-@Requires(beans = JsonConfiguration.class)
-@Requires(classes = JsonView.class)
-@Requires(property = JsonViewServerFilter.PROPERTY_JSON_VIEW_ENABLED)
-@ServerFilter("/**")
-@Internal
-public class JsonViewServerFilter implements Ordered {
```
Comment (graemerocher) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java:57]: strictly speaking we have to leave this in place and deprecate it

**Review by graemerocher — APPROVED** [review]

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java (hunk 7 above)
Comment (dstepanov) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java:57]: Why? It's internal.

Comment (graemerocher) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java:57]: oh in that case no problem :)

**Pull request merged**